File: cpp/src/TravelTimeData.cpp
#include <TravelTimeData.h>

#include <string>
#include <limits>
#include <exception>
#include <vector>

// JSON Keys
#define TYPE_KEY "Type"
#define PHASE_KEY "Phase"
#define TRAVELTIME_KEY "TravelTime"
#define DISTANCEDERIVATIVE_KEY "DistanceDerivative"
#define DEPTHDERIVATIVE_KEY "DepthDerivative"
#define RAYDERIVATIVE_KEY "RayDerivative"
#define STATISTICALSPREAD_KEY "StatisticalSpread"
#define OBSERVABILITY_KEY "Observability"
#define TELESEISMICPHASEGROUP_KEY "TeleseismicPhaseGroup"
#define AUXILIARYPHASEGROUP_KEY "AuxiliaryPhaseGroup"
#define LOCATIONUSEFLAG_KEY "LocationUseFlag"
#define ASSOCIATIONWEIGHTFLAG_KEY "AssociationWeightFlag"

namespace processingformats {

TravelTimeData::TravelTimeData() {
	type = "TTData";
	phase = "";
	travelTime = std::numeric_limits<double>::quiet_NaN();
	distanceDerivative = std::numeric_limits<double>::quiet_NaN();
	depthDerivative = std::numeric_limits<double>::quiet_NaN();
	rayDerivative = std::numeric_limits<double>::quiet_NaN();
	statisticalSpread = std::numeric_limits<double>::quiet_NaN();
	observability = std::numeric_limits<double>::quiet_NaN();
	teleseismicPhaseGroup = "";
	auxiliaryPhaseGroup = "";
	locationUseFlag = false;
	associationWeightFlag = false;
}

TravelTimeData::TravelTimeData(std::string newPhase, double newTravelTime,
								double newDistanceDerivative,
								double newDepthDerivative,
								double newRayDerivative,
								double newStatisticalSpread,
								double newObservability,
								std::string newTeleseismicPhaseGroup,
								std::string newAuxiliaryPhaseGroup,
								bool newLocationUseFlag,
								bool newAssociationWeightFlag) {
	type = "TTData";
	phase = newPhase;
	travelTime = newTravelTime;
	distanceDerivative = newDistanceDerivative;
	depthDerivative = newDepthDerivative;
	rayDerivative = newRayDerivative;
	statisticalSpread = newStatisticalSpread;
	observability = newObservability;
	teleseismicPhaseGroup = newTeleseismicPhaseGroup;
	auxiliaryPhaseGroup = newAuxiliaryPhaseGroup;
	locationUseFlag = newLocationUseFlag;
	associationWeightFlag = newAssociationWeightFlag;
}

TravelTimeData::TravelTimeData(rapidjson::Value &json) {
	// required values
	// type
	if ((json.HasMember(TYPE_KEY) == true)
			&& (json[TYPE_KEY].IsString() == true)) {
		type = std::string(json[TYPE_KEY].GetString(),
							json[TYPE_KEY].GetStringLength());
	} else {
		type = "";
	}

	// phase
	if ((json.HasMember(PHASE_KEY) == true)
			&& (json[PHASE_KEY].IsString() == true)) {
		phase = std::string(json[PHASE_KEY].GetString(),
							json[PHASE_KEY].GetStringLength());
	} else {
		phase = "";
	}

	// travelTime
	if ((json.HasMember(TRAVELTIME_KEY) == true)
			&& (json[TRAVELTIME_KEY].IsNumber() == true)
			&& (json[TRAVELTIME_KEY].IsDouble() == true))
		travelTime = json[TRAVELTIME_KEY].GetDouble();
	else
		travelTime = std::numeric_limits<double>::quiet_NaN();

	// distanceDerivative
	if ((json.HasMember(DISTANCEDERIVATIVE_KEY) == true)
			&& (json[DISTANCEDERIVATIVE_KEY].IsNumber() == true)
			&& (json[DISTANCEDERIVATIVE_KEY].IsDouble() == true))
		distanceDerivative = json[DISTANCEDERIVATIVE_KEY].GetDouble();
	else
		distanceDerivative = std::numeric_limits<double>::quiet_NaN();

	// depthDerivative
	if ((json.HasMember(DEPTHDERIVATIVE_KEY) == true)
			&& (json[DEPTHDERIVATIVE_KEY].IsNumber() == true)
			&& (json[DEPTHDERIVATIVE_KEY].IsDouble() == true))
		depthDerivative = json[DEPTHDERIVATIVE_KEY].GetDouble();
	else
		depthDerivative = std::numeric_limits<double>::quiet_NaN();

	// rayDerivative
	if ((json.HasMember(RAYDERIVATIVE_KEY) == true)
			&& (json[RAYDERIVATIVE_KEY].IsNumber() == true)
			&& (json[RAYDERIVATIVE_KEY].IsDouble() == true))
		rayDerivative = json[RAYDERIVATIVE_KEY].GetDouble();
	else
		rayDerivative = std::numeric_limits<double>::quiet_NaN();

	// statisticalSpread
	if ((json.HasMember(STATISTICALSPREAD_KEY) == true)
			&& (json[STATISTICALSPREAD_KEY].IsNumber() == true)
			&& (json[STATISTICALSPREAD_KEY].IsDouble() == true))
		statisticalSpread = json[STATISTICALSPREAD_KEY].GetDouble();
	else
		statisticalSpread = std::numeric_limits<double>::quiet_NaN();

	// observability
	if ((json.HasMember(OBSERVABILITY_KEY) == true)
			&& (json[OBSERVABILITY_KEY].IsNumber() == true)
			&& (json[OBSERVABILITY_KEY].IsDouble() == true))
		observability = json[OBSERVABILITY_KEY].GetDouble();
	else
		observability = std::numeric_limits<double>::quiet_NaN();

	// teleseismicPhaseGroup
	if ((json.HasMember(TELESEISMICPHASEGROUP_KEY) == true)
			&& (json[TELESEISMICPHASEGROUP_KEY].IsString() == true))
		teleseismicPhaseGroup = std::string(
				json[TELESEISMICPHASEGROUP_KEY].GetString(),
				json[TELESEISMICPHASEGROUP_KEY].GetStringLength());
	else
		teleseismicPhaseGroup = "";

	// auxiliaryPhaseGroup
	if ((json.HasMember(AUXILIARYPHASEGROUP_KEY) == true)
			&& (json[AUXILIARYPHASEGROUP_KEY].IsString() == true))
		auxiliaryPhaseGroup = std::string(
				json[AUXILIARYPHASEGROUP_KEY].GetString(),
				json[AUXILIARYPHASEGROUP_KEY].GetStringLength());
	else
		auxiliaryPhaseGroup = "";

	// locationUseFlag
	if ((json.HasMember(LOCATIONUSEFLAG_KEY) == true)
			&& (json[LOCATIONUSEFLAG_KEY].IsBool() == true)) {
		locationUseFlag = json[LOCATIONUSEFLAG_KEY].GetBool();
	} else {
		locationUseFlag = true;
	}

	// associationWeightFlag
	if ((json.HasMember(ASSOCIATIONWEIGHTFLAG_KEY) == true)
			&& (json[ASSOCIATIONWEIGHTFLAG_KEY].IsBool() == true)) {
		associationWeightFlag = json[ASSOCIATIONWEIGHTFLAG_KEY].GetBool();
	} else {
		associationWeightFlag = true;
	}
}

TravelTimeData::TravelTimeData(const TravelTimeData & newTravelTimeData) {
	type = "TTData";
	phase = newTravelTimeData.phase;
	travelTime = newTravelTimeData.travelTime;
	distanceDerivative = newTravelTimeData.distanceDerivative;
	depthDerivative = newTravelTimeData.depthDerivative;
	rayDerivative = newTravelTimeData.rayDerivative;
	statisticalSpread = newTravelTimeData.statisticalSpread;
	observability = newTravelTimeData.observability;
	teleseismicPhaseGroup = newTravelTimeData.teleseismicPhaseGroup;
	auxiliaryPhaseGroup = newTravelTimeData.auxiliaryPhaseGroup;
	locationUseFlag = newTravelTimeData.locationUseFlag;
	associationWeightFlag = newTravelTimeData.associationWeightFlag;
}

TravelTimeData::~TravelTimeData() {
}

rapidjson::Value & TravelTimeData::toJSON(
		rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// type
	rapidjson::Value typevalue;
	typevalue.SetString(rapidjson::StringRef(type.c_str()), allocator);
	json.AddMember(TYPE_KEY, typevalue, allocator);

	// phase
	if (phase != "") {
		rapidjson::Value phasevalue;
		phasevalue.SetString(rapidjson::StringRef(phase.c_str()), allocator);
		json.AddMember(PHASE_KEY, phasevalue, allocator);
	}

	// travelTime
	if (std::isnan(travelTime) != true)
		json.AddMember(TRAVELTIME_KEY, travelTime, allocator);

	// distanceDerivative
	if (std::isnan(distanceDerivative) != true)
		json.AddMember(DISTANCEDERIVATIVE_KEY, distanceDerivative, allocator);

	// depthDerivative
	if (std::isnan(depthDerivative) != true)
		json.AddMember(DEPTHDERIVATIVE_KEY, depthDerivative, allocator);

	// rayDerivative
	if (std::isnan(rayDerivative) != true)
		json.AddMember(RAYDERIVATIVE_KEY, rayDerivative, allocator);

	// statisticalSpread
	if (std::isnan(statisticalSpread) != true)
		json.AddMember(STATISTICALSPREAD_KEY, statisticalSpread, allocator);

	// observability
	if (std::isnan(observability) != true)
		json.AddMember(OBSERVABILITY_KEY, observability, allocator);

	// teleseismicPhaseGroup
	if (teleseismicPhaseGroup != "") {
		rapidjson::Value televalue;
		televalue.SetString(rapidjson::StringRef(teleseismicPhaseGroup.c_str()),
							allocator);
		json.AddMember(TELESEISMICPHASEGROUP_KEY, televalue, allocator);
	}

	// snr
	if (auxiliaryPhaseGroup != "") {
		rapidjson::Value auxvalue;
		auxvalue.SetString(rapidjson::StringRef(auxiliaryPhaseGroup.c_str()),
							allocator);
		json.AddMember(AUXILIARYPHASEGROUP_KEY, auxvalue, allocator);
	}

	// locationUseFlag
	json.AddMember(LOCATIONUSEFLAG_KEY, locationUseFlag, allocator);

	// associationWeightFlag
	json.AddMember(ASSOCIATIONWEIGHTFLAG_KEY, associationWeightFlag, allocator);

	return (json);
}

std::vector<std::string> TravelTimeData::getErrors() {
	std::vector<std::string> errorlist;

	// required data
	// type
	if (type == "") {
		// type empty
		errorlist.push_back("Empty Type in TravelTimeData Class.");
	} else if (type != "TTData") {
		// wrong type
		errorlist.push_back("Non-TTData type in TravelTimeData Class.");
	}

	// phase
	if (phase == "") {
		// phase empty
		errorlist.push_back("Empty Phase in TravelTimeData Class.");
	}

	// travel time
	if (std::isnan(travelTime) != true) {
		// travel time not found
		errorlist.push_back("No Travel Time in TravelTimeData Class.");
	}

	// distance derivative
	if (std::isnan(distanceDerivative) != true) {
		// distance derivative not found
		errorlist.push_back("No Distance Derivative in TravelTimeData Class.");
	}

	// depth derivative
	if (std::isnan(depthDerivative) != true) {
		// depth derivative not found
		errorlist.push_back("No Depth Derivative in TravelTimeData Class.");
	}

	// ray derivative
	if (std::isnan(rayDerivative) != true) {
		// ray derivative not found
		errorlist.push_back("No Ray Derivative in TravelTimeData Class.");
	}

	// statistical spread
	if (std::isnan(statisticalSpread) != true) {
		// statistical spread not found
		errorlist.push_back("No Statistical Spread in TravelTimeData Class.");
	}

	// observability
	if (std::isnan(observability) != true) {
		// observability not found
		errorlist.push_back("No Observability in TravelTimeData Class.");
	}

	// teleseismic phase group
	if (teleseismicPhaseGroup == "") {
		// teleseismic phase group not found
		errorlist.push_back(
				"No Teleseismic Phase Group in TravelTimeData Class.");
	}

	// auxiliary phase group
	if (auxiliaryPhaseGroup == "") {
		// auxiliary phase group not found
		errorlist.push_back(
				"No Auxiliary Phase Group in TravelTimeData Class.");
	}

	return (errorlist);
}
}  // namespace processingformats

File: java/src/main/java/gov/usgs/processingformats/TravelTimeData.java
package gov.usgs.processingformats;

import java.util.*;
import org.json.simple.JSONObject;

/**
 * a conversion class used to create, parse, and validate travel time request data
 *
 * @author U.S. Geological Survey &lt;jpatton at usgs.gov&gt;
 */
public class TravelTimeData implements ProcessingInt {

  /** JSON Keys */
  public static final String PHASE_KEY = "Phase";

  public static final String TRAVELTIME_KEY = "TravelTime";
  public static final String DISTANCEDERIVATIVE_KEY = "DistanceDerivative";
  public static final String DEPTHDERIVATIVE_KEY = "DepthDerivative";
  public static final String RAYDERIVATIVE_KEY = "RayDerivative";
  public static final String STATISTICALSPREAD_KEY = "StatisticalSpread";
  public static final String OBSERVABILITY_KEY = "Observability";
  public static final String TELESEISMICPHASEGROUP_KEY = "TeleseismicPhaseGroup";
  public static final String AUXILIARYPHASEGROUP_KEY = "AuxiliaryPhaseGroup";
  public static final String LOCATIONUSEFLAG_KEY = "LocationUseFlag";
  public static final String ASSOCIATIONWEIGHTFLAG_KEY = "AssociationWeightFlag";

  /** Required seismic Phase code */
  public String Phase;

  /** Required travel time in seconds */
  public Double TravelTime;

  /** Required derivative with respect to distance of the travel time in seconds/degree. */
  public Double DistanceDerivative;

  /** Required derivative with respect to depth of the travel time in seconds/kilometer. */
  public Double DepthDerivative;

  /** Required derivative with respect to ray parameter of the travel time in degrees/second. */
  public Double RayDerivative;

  /** Required observed travel time scatter in seconds. */
  public Double StatisticalSpread;

  /** Required statistical Observability of the seismic Phase. */
  public Double Observability;

  /** Required teleseismic Phase group identifier. */
  public String TeleseismicPhaseGroup;

  /** Required auxiliary Phase group identifier. */
  public String AuxiliaryPhaseGroup;

  /** Required flag indicating whether the Phase may be used in a location. */
  public Boolean LocationUseFlag;

  /** Required flag indicating whether the Phase should be down weighted in association. */
  public Boolean AssociationWeightFlag;

  /** The constructor for the TravelTimeData class. Initializes members to null values. */
  public TravelTimeData() {

    reload(null, null, null, null, null, null, null, null, null, null, null);
  }

  /**
   * Advanced constructor
   *
   * <p>The advanced constructor for the TravelTimeData class. Initializes members to provided
   * values.
   *
   * @param newPhase - A String containing the seismic Phase code
   * @param newTravelTime - A Double containing the travel time in seconds
   * @param newDistanceDerivative - A Double containing the derivative with respect to distance of
   *     the travel time in seconds/degree
   * @param newDepthDerivative - A Double containing the derivative with respect to ray parameter of
   *     the travel time in degrees/second
   * @param newRayDerivative - A Double containing the derivative with respect to ray parameter of
   *     the travel time in degrees/second
   * @param newStatisticalSpread - A Double containing the observed travel time scatter in seconds
   * @param newObservability - A Double containing the statistical Observability of the seismic
   *     Phase
   * @param newTeleseismicPhaseGroup - A String containing the teleseismic Phase group identifier
   * @param newAuxiliaryPhaseGroup - A String containing the auxiliary Phase group identifier
   * @param newLocationUseFlag - A Boolean containing the flag indicating whether the Phase may be
   *     used in a location
   * @param newAssociationWeightFlag - A Boolean containing the flag indicating whether the Phase
   *     should be down weighted in assocation
   */
  public TravelTimeData(
      String newPhase,
      Double newTravelTime,
      Double newDistanceDerivative,
      Double newDepthDerivative,
      Double newRayDerivative,
      Double newStatisticalSpread,
      Double newObservability,
      String newTeleseismicPhaseGroup,
      String newAuxiliaryPhaseGroup,
      Boolean newLocationUseFlag,
      Boolean newAssociationWeightFlag) {

    reload(
        newPhase,
        newTravelTime,
        newDistanceDerivative,
        newDepthDerivative,
        newRayDerivative,
        newStatisticalSpread,
        newObservability,
        newTeleseismicPhaseGroup,
        newAuxiliaryPhaseGroup,
        newLocationUseFlag,
        newAssociationWeightFlag);
  }

  /**
   * Constructs the class from a JSONObject, populating members
   *
   * @param newJSONObject - A JSONObject.
   */
  public TravelTimeData(JSONObject newJSONObject) {

    // Required values
    // Phase
    if (newJSONObject.containsKey(PHASE_KEY)) {
      Phase = newJSONObject.get(PHASE_KEY).toString();
    } else {
      Phase = null;
    }

    // travel time
    if (newJSONObject.containsKey(TRAVELTIME_KEY)) {
      TravelTime = (double) newJSONObject.get(TRAVELTIME_KEY);
    } else {
      TravelTime = null;
    }

    // distance Derivative
    if (newJSONObject.containsKey(DISTANCEDERIVATIVE_KEY)) {
      DistanceDerivative = (double) newJSONObject.get(DISTANCEDERIVATIVE_KEY);
    } else {
      DistanceDerivative = null;
    }

    // depth Derivative
    if (newJSONObject.containsKey(DEPTHDERIVATIVE_KEY)) {
      DepthDerivative = (double) newJSONObject.get(DEPTHDERIVATIVE_KEY);
    } else {
      DepthDerivative = null;
    }

    // ray Derivative
    if (newJSONObject.containsKey(RAYDERIVATIVE_KEY)) {
      RayDerivative = (double) newJSONObject.get(RAYDERIVATIVE_KEY);
    } else {
      RayDerivative = null;
    }

    // statistical spread
    if (newJSONObject.containsKey(STATISTICALSPREAD_KEY)) {
      StatisticalSpread = (double) newJSONObject.get(STATISTICALSPREAD_KEY);
    } else {
      StatisticalSpread = null;
    }

    // Observability
    if (newJSONObject.containsKey(OBSERVABILITY_KEY)) {
      Observability = (double) newJSONObject.get(OBSERVABILITY_KEY);
    } else {
      Observability = null;
    }

    // teleseismic Phase group
    if (newJSONObject.containsKey(TELESEISMICPHASEGROUP_KEY)) {
      TeleseismicPhaseGroup = newJSONObject.get(TELESEISMICPHASEGROUP_KEY).toString();
    } else {
      TeleseismicPhaseGroup = null;
    }

    // auxiliary Phase group
    if (newJSONObject.containsKey(AUXILIARYPHASEGROUP_KEY)) {
      AuxiliaryPhaseGroup = newJSONObject.get(AUXILIARYPHASEGROUP_KEY).toString();
    } else {
      AuxiliaryPhaseGroup = null;
    }

    // location use flag
    if (newJSONObject.containsKey(LOCATIONUSEFLAG_KEY)) {
      LocationUseFlag = (boolean) newJSONObject.get(LOCATIONUSEFLAG_KEY);
    } else {
      LocationUseFlag = null;
    }

    // association weight flag
    if (newJSONObject.containsKey(ASSOCIATIONWEIGHTFLAG_KEY)) {
      AssociationWeightFlag = (boolean) newJSONObject.get(ASSOCIATIONWEIGHTFLAG_KEY);
    } else {
      AssociationWeightFlag = null;
    }
  }

  /**
   * Constructs the class from a TravelTimeData object, populating members (copy constructor)
   *
   * @param sourceObject - A TravelTimeData object.
   */
  public TravelTimeData(TravelTimeData sourceObject) {
    reload(
        sourceObject.Phase,
        sourceObject.TravelTime,
        sourceObject.DistanceDerivative,
        sourceObject.DepthDerivative,
        sourceObject.RayDerivative,
        sourceObject.StatisticalSpread,
        sourceObject.Observability,
        sourceObject.TeleseismicPhaseGroup,
        sourceObject.AuxiliaryPhaseGroup,
        sourceObject.LocationUseFlag,
        sourceObject.AssociationWeightFlag);
  }

  /**
   * Reload Function
   *
   * <p>The reload function for the TravelTimeData class. Initializes members to provided values.
   *
   * @param newPhase - A String containing the seismic Phase code
   * @param newTravelTime - A Double containing the travel time in seconds
   * @param newDistanceDerivative - A Double containing the derivative with respect to distance of
   *     the travel time in seconds/degree
   * @param newDepthDerivative - A Double containing the derivative with respect to ray parameter of
   *     the travel time in degrees/second
   * @param newRayDerivative - A Double containing the derivative with respect to ray parameter of
   *     the travel time in degrees/second
   * @param newStatisticalSpread - A Double containing the observed travel time scatter in seconds
   * @param newObservability - A Double containing the statistical Observability of the seismic
   *     Phase
   * @param newTeleseismicPhaseGroup - A String containing the teleseismic Phase group identifier
   * @param newAuxiliaryPhaseGroup - A String containing the auxiliary Phase group identifier
   * @param newLocationUseFlag - A Boolean containing the flag indicating whether the Phase may be
   *     used in a location
   * @param newAssociationWeightFlag - A Boolean containing the flag indicating whether the Phase
   *     should be down weighted in association
   */
  public void reload(
      String newPhase,
      Double newTravelTime,
      Double newDistanceDerivative,
      Double newDepthDerivative,
      Double newRayDerivative,
      Double newStatisticalSpread,
      Double newObservability,
      String newTeleseismicPhaseGroup,
      String newAuxiliaryPhaseGroup,
      Boolean newLocationUseFlag,
      Boolean newAssociationWeightFlag) {

    Phase = newPhase;
    TravelTime = newTravelTime;
    DistanceDerivative = newDistanceDerivative;
    DepthDerivative = newDepthDerivative;
    RayDerivative = newRayDerivative;
    StatisticalSpread = newStatisticalSpread;
    Observability = newObservability;
    TeleseismicPhaseGroup = newTeleseismicPhaseGroup;
    AuxiliaryPhaseGroup = newAuxiliaryPhaseGroup;
    LocationUseFlag = newLocationUseFlag;
    AssociationWeightFlag = newAssociationWeightFlag;
  }

  /**
   * Converts the contents of the class to a json object
   *
   * @return Returns a JSONObject containing the class contents
   */
  @SuppressWarnings("unchecked")
  public JSONObject toJSON() {

    JSONObject newJSONObject = new JSONObject();

    // Phase
    if (Phase != null) {
      newJSONObject.put(PHASE_KEY, Phase);
    }

    // travel time
    if (TravelTime != null) {
      newJSONObject.put(TRAVELTIME_KEY, TravelTime);
    }

    // distance Derivative
    if (DistanceDerivative != null) {
      newJSONObject.put(DISTANCEDERIVATIVE_KEY, DistanceDerivative);
    }

    // depth Derivative
    if (DepthDerivative != null) {
      newJSONObject.put(DEPTHDERIVATIVE_KEY, DepthDerivative);
    }

    // ray Derivative
    if (RayDerivative != null) {
      newJSONObject.put(RAYDERIVATIVE_KEY, RayDerivative);
    }

    // statistical spread
    if (StatisticalSpread != null) {
      newJSONObject.put(STATISTICALSPREAD_KEY, StatisticalSpread);
    }

    // Observability
    if (Observability != null) {
      newJSONObject.put(OBSERVABILITY_KEY, Observability);
    }

    // teleseismic Phase group
    if (TeleseismicPhaseGroup != null) {
      newJSONObject.put(TELESEISMICPHASEGROUP_KEY, TeleseismicPhaseGroup);
    }

    // auxiliary Phase group
    if (AuxiliaryPhaseGroup != null) {
      newJSONObject.put(AUXILIARYPHASEGROUP_KEY, AuxiliaryPhaseGroup);
    }

    // location use flag
    if (LocationUseFlag != null) {
      newJSONObject.put(LOCATIONUSEFLAG_KEY, LocationUseFlag);
    }

    // association use flag
    if (AssociationWeightFlag != null) {
      newJSONObject.put(ASSOCIATIONWEIGHTFLAG_KEY, AssociationWeightFlag);
    }

    return (newJSONObject);
  }

  /**
   * Validates the class.
   *
   * @return Returns true if successful
   */
  public boolean isValid() {
    if (getErrors() == null) {
      return (true);
    } else if (getErrors().size() == 0) {
      return (true);
    } else {
      return (false);
    }
  }

  /**
   * Gets any validation errors in the class.
   *
   * @return Returns a List&lt;String&gt; of any errors found
   */
  public ArrayList<String> getErrors() {
    ArrayList<String> errorList = new ArrayList<String>();

    if (Phase == null) {
      // Phase not found
      errorList.add("No Phase in TravelTimeData Class.");
    } else if (Phase.isEmpty()) {
      // Phase empty
      errorList.add("Empty Phase in TravelTimeData Class.");
    }

    // travel time
    if (TravelTime == null) {
      // travel time not found
      errorList.add("No Travel Time in TravelTimeData Class.");
    }

    // distance derivative
    if (DistanceDerivative == null) {
      // distance derivative not found
      errorList.add("No Distance Derivative in TravelTimeData Class.");
    }

    // depth derivative
    if (DepthDerivative == null) {
      // depth derivative not found
      errorList.add("No Depth Derivative in TravelTimeData Class.");
    }

    // ray derivative
    if (RayDerivative == null) {
      // ray derivative not found
      errorList.add("No Ray Derivative in TravelTimeData Class.");
    }

    // statistical spread
    if (StatisticalSpread == null) {
      // statistical spread not found
      errorList.add("No Statistical Spread in TravelTimeData Class.");
    }

    // Observability
    if (Observability == null) {
      // Observability not found
      errorList.add("No Observability in TravelTimeData Class.");
    }

    // teleseismic Phase group
    if (TeleseismicPhaseGroup == null) {
      // teleseismic Phase group not found
      errorList.add("No Teleseismic Phase Group in TravelTimeData Class.");
    }

    // auxiliary Phase group
    if (AuxiliaryPhaseGroup == null) {
      // auxiliary Phase group not found
      errorList.add("No Auxiliary Phase Group in TravelTimeData Class.");
    }

    // location use flag
    if (LocationUseFlag == null) {
      // location use flag not found
      errorList.add("No Location Use Flag in TravelTimeData Class.");
    }

    // association weight flag
    if (AssociationWeightFlag == null) {
      // association weight flag not found
      errorList.add("No Association Weight Flag in TravelTimeData Class.");
    }

    return (errorList);
  }
}
